packages/ui/src/components/PropertiesModal/PropertiesModal.tsx

```diff
@@ -1,33 +1,141 @@
 import './PropertiesModal.scss';
+import { PropertiesTabs } from './PropertiesTabs';
+import { transformPropertiesIntoTab } from './camel-properties-to-tab';
 
 interface IPropertiesModalProps {
   tile: ITile;
   onClose: () => void;
   isModalOpen: boolean;
 }
 
+const transformCamelComponentIntoTab = (
+  tabsToRender: IPropertiesTab[],
+  componentDef: ICamelComponentDefinition,
+): void => {
+  let tab = transformPropertiesIntoTab(
+    [
+      {
+        transformFunction: camelComponentPropertiesToTable,
+        propertiesObject: componentDef.componentProperties,
+      },
+    ],
+    'Component Options',
+  );
+  if (tab) tabsToRender.push(tab);
+
+  // properties, contains 2 subtables divided according to Kind
+  tab = transformPropertiesIntoTab(
+    [
+      {
+        transformFunction: camelComponentPropertiesToTable,
+        propertiesObject: componentDef.properties,
+        filter: {
+          filterKey: 'kind',
+          filterValue: 'path',
+        },
+        tableCaption: 'path parameters',
+      },
+      {
+        transformFunction: camelComponentPropertiesToTable,
+        propertiesObject: componentDef.properties,
+        filter: {
+          filterKey: 'kind',
+          filterValue: 'parameter',
+        },
+        tableCaption: 'query parameters',
+      },
+    ],
+    'Endpoint Options',
+  );
+  if (tab) tabsToRender.push(tab);
+
+  // headers, only if exists
+  if (componentDef.headers) {
+    let tab = transformPropertiesIntoTab(
+      [
+        {
+          transformFunction: camelComponentPropertiesToTable,
+          propertiesObject: componentDef.headers,
+        },
+      ],
+      'Message Headers',
+    );
+    if (tab) tabsToRender.push(tab);
+  }
+
+  // apis, only if exists
+  if (componentDef.apis) {
+    let tab = transformPropertiesIntoTab(
+      [
+        {
+          transformFunction: camelComponentApisToTable,
+          propertiesObject: componentDef.apis,
+        },
+      ],
+      'APIs',
+    );
+    if (tab) tabsToRender.push(tab);
+  }
+};
+
+const transformCamelProcessorComponentIntoTab = (
+  tabsToRender: IPropertiesTab[],
+  processorDef: ICamelProcessorDefinition,
+): void => {
+  let tab = transformPropertiesIntoTab(
+    [
+      {
+        transformFunction: camelProcessorPropertiesToTable,
+        propertiesObject: processorDef.properties,
+      },
+    ],
+    'Options',
+  );
+  if (tab) tabsToRender.push(tab);
+};
+
+const transformKameletComponentIntoTab = (tabsToRender: IPropertiesTab[], kameletDef: IKameletDefinition): void => {
+  let tab = transformPropertiesIntoTab(
+    [
+      {
+        transformFunction: kameletToPropertiesTable,
+        propertiesObject: kameletDef,
+      },
+    ],
+    'Options',
+  );
+  if (tab) tabsToRender.push(tab);
+};
+
 export const PropertiesModal: FunctionComponent<IPropertiesModalProps> = (props) => {
-  let table: IPropertiesTable;
+  let tabs: IPropertiesTab[] = [];
   switch (props.tile.type) {
```

Comment:
The thing here is that it will make the calculations on every render. It would be more convenient to store the `tabs` object using a `useState` hook and then update it using a `useEffect` that depends on `props.tile.type`.

The idea would be more or less something like this:

```typescript
const [tabs, setTabs] = useState<IPropertiesTab[]>([]);

useEffect(() => {
  //  switch logic goes here

  // store the resulting calculation like
  setTabs(CALCULATED_TABS_HERE);
}, [props.tile.type]);
```

Also, is worth mentioning that unless we plan to reuse the returning array, it would be more convenient (IMHO) to return an array from the function, rather than providing one from the outside.

```typescript
let tabs: IPropertiesTab[] = [];
transformCamelComponentIntoTab(tabs, props.tile.rawObject as ICamelComponentDefinition);

// vs 

const tabs = transformCamelComponentIntoTab(props.tile.rawObject as ICamelComponentDefinition);
```

Reply:
Tabs are in a different component, there is no re-render in the modal component However, of course, it can be some in the future so I will move it into useEffect() .

Reply:
@lordrip  Isn't there better to use `useCallback` / `useMemo` hooks in this case?
Because `useEffect` is running only after the component is rendered, so, for the first time, the empty table is shown for a little millisecond (because the table is null for first rendering). 
In other words, in this case, the computation is needed before the first rendering.
e.g.
```
let tabs = useMemo(() => {
    switch (props.tile.type) {
      ...
    }
  }, [props.tile]);
```

Reply:
You're right, let's go with your approach then :muscle: